### DISCUSSION TITLE
Improved support for Gemini monitoring

### DISCUSSION THREAD

**Discussion by bdagnino** [langfuse/langfuse#4777] [Ideas]
### Describe the feature or potential improvement

I'd like to use langfuse to monitor an app that uses the Gemini API. The documentation I could find is based on Vertex, but I'm using the [Google AI Python SDK for the Gemini API](https://pypi.org/project/google-generativeai/). Ideally I'd like to be as easy as it is with OpenAI, where you just import the Langfuse client and go from there (which is awesome DX!). 

Hopes this makes sense but happy to elaborate of be pointed in the right direction if this already exists! 

Thanks!

**Comment by marcklingen** [comment]
Thanks for sharing! Totally makes sense. Let's track this here.

In the meantime, you could wrap the gemini sdk as shown [here](https://langfuse.com/docs/sdk/python/decorators#log-any-llm-call) or use one of the framework integration (like [LangChain](https://langfuse.com/docs/integrations/langchain/tracing)) to use gemini with automated instrumentation in Langfuse. Is one of these a viable option?

**Comment by bdagnino** [comment]
Awesome, thanks for looking into this. I believe Gemini's DX will be super important as more people adopt it so hopefully this helps other as well. 

About your suggests. We don't use Langchain/Llamaindex so that's not a good fit. 

I tried already wrapping the gemini sdk, and it works for the basic use case of logging the input and output, but if you want to do other things like token count it's not so straightforward. E.g. we use Instructor and that conflicts with the way you get the token count from gemini's response. 

There are ways around it, but that's what made my experience with Langfuse great when working with OpenAI: you don't have to do the workarounds 😅

**Reply by marcklingen** [reply]
Totally agree. Unfortunately this is the only way to get accurate token counts as we cannot maintain these wrapped sdks for all providers (a lot of effort as the sdks change over time).

Learn more about the instrumentation for cost tracking here: https://langfuse.com/docs/model-usage-and-cost

**Reply by marcklingen** [reply]
Let me know if you can't figure this out. It would be helpful if you could share what worked for you in the end. I'd love to add it to the documentation page here: https://langfuse.com/docs/integrations/google-vertex-ai

**Comment by FionnD** [comment]
As also mentioned [here](https://github.com/orgs/langfuse/discussions/5182), it would be great if someone could share how to track token usage with Gemini.
